src/cp/cp.rs

```diff
@@ -658,27 +686,94 @@ fn copy(sources: &[Source], target: &Target, options: &Options) -> CopyResult<()
     let target_type = TargetType::determine(sources, target);
     verify_target_type(target, &target_type)?;
 
+    let mut preserve_hard_links = false;
+    for attribute in &options.preserve_attributes {
+        if *attribute == Attribute::Links {
+            preserve_hard_links = true;
+        }
+    }
+    #[cfg(unix)]
+    let mut hard_links: Vec<(String, u64)> = vec![];
+    
+    #[cfg(windows)]
+    let mut hard_links: Vec<(String, (u32, u32))> = vec![];
+
     let mut non_fatal_errors = false;
     let mut seen_sources = HashSet::with_capacity(sources.len());
-
     for source in sources {
         if seen_sources.contains(source) {
             show_warning!("source '{}' specified more than once", source.display());
 
-        } else if let Err(error) = copy_source(source, target, &target_type, options) {
-            show_error!("{}", error);
-            match error {
-                Error::Skipped(_) => (),
-                _ => non_fatal_errors = true,
+        } else {
+            let mut found_hard_link = false;
+            if preserve_hard_links {
+                #[cfg(unix)]
```

Comment:
I'd rather this `#[cfg(unix)]` section and the `#[cfg(windows)]` section below be split off into separate functions.